docs/processing.md

```diff
@@ -0,0 +1,288 @@
+`HistogramDataPoint`, with the information from higher levels like `Metric` or the data type available as virtual fields.
+
+Virtual fields for all signals: `resource`, `library_info`.
+Virtual fields for metrics: `descriptor`, which contains `metric.name`, `metric.description`, `metric.unit`.
+
+Navigation can then be used with a simple expression language for identifying telemetry to operate on.
+
+`... where name = "GET /cats"`
+`... from span where attributes["http.target"] = "/health"`
+`... where resource.attributes["deployment"] = "canary"`
+`... from metric where descriptor.metric_type = gauge`
+`... from metric where descriptor.metric_name = "http.active_requests"`
+
+Fields should always be fully specified - for example `attributes` refers to the `attributes` field in the telemetry, not
+the `resource`. In the future, we may allow shorthand for accessing scoped information that is not ambiguous.
+
+Having selected telemetry to operate on, any needed operations can be defined as functions. Known useful functions should
+be implemented within the collector itself, provide registration from extension modules to allow customization with
+contrib components, and in the future can even allow user plugins possibly through WASM, similar to work in 
+[HTTP proxies](https://github.com/proxy-wasm/spec). The arguments to operations will primarily be field expressions,
+allowing the operation to mutate telemetry as needed.
+
+### Examples
+
+Remove a forbidden attribute such as `http.request.header.authorization` from spans only
+
+`delete(attributes["http.request.header.authorization"]) from traces`
+
+Remove all attributes except for some
+
+`keep(attributes, "http.method", "http.status_code") from metrics`
+
+Reduce cardinality of an attribute
+
+`replace_wildcards("/user/*/list/*", "/user/{userId}/list/{listId}", attributes["http.target"]) from traces`
+
+Reduce cardinality of a span name
+
+`replace_wildcards("GET /user/*/list/*", "GET /user/{userId}/list/{listId}", name) from traces`
+
+Decrease the size of the telemetry payload by removing large resource attributes
+
+`delete(resource.attributes["process.command_line"]) from traces`
+
+Filtering out signals such as by removing all metrics with a `http.target` of `/health`
+
+`drop() where attributes["http.target"] = "/health" from metrics`
+
+Attach information from resource into telemetry, for example adding certain resource fields as metric attributes
+
+`set(attributes["k8s_pod"], resource.attributes["k8s.pod.name"]) from metrics`
+
+Stateful processing can also be modeled by the language. The processor implementation would set up the state while
+parsing the configuration.
+
+Create duration_metric with two attributes copied from a span
+
+```
+create_histogram("duration", end_time_nanos - start_time_nanos) from traces
+keep(attributes, "http.method") from metrics where descriptor.metric_name = "duration"
```

Comment:
Sorry I had meant to remove these cases but forgot this one. When originally writing the doc I forgot about the one signal-per-pipeline limitation.

In the future, I think shared pipelines could be cool but not for now, especially because collector core has no way of supporting that